New protocol implementation

amshinde opened the pull request.

dlespiau commented on src/utils.c:
```diff
@@ -97,6 +97,32 @@ set_fd_nonblocking(int fd)
 }
 
 /*!
+ * Store short integer as big endian in buffer
```
why not use C library functions here? ntohl, ntohs, htons, htonl

dlespiau commented on src/shim.c:
```diff
@@ -249,6 +249,98 @@ send_proxy_hyper_message(int fd, const char *hyper_cmd, const char *json) {
+		return false;
+	}
+
+	while (offset < len) {
```
s/len/total_len/

amshinde replied:
Fixed.

jodh-intel commented on src/shim.h:
```diff
@@ -27,6 +27,59 @@ struct cc_shim {
+};
+
+enum notificationtype {
+	//Assuming that the exit code
```
partial comment?

amshinde replied:
Fixed.

jodh-intel commented on src/shim.c:
```diff
@@ -43,9 +48,8 @@ struct pollfd poll_fds[MAX_POLL_FDS] = {{-1}};
-#define STDIN_INDEX 3
+#define PROXY_SOCK_INDEX 1
+#define STDIN_INDEX 2
 
```
Not a blocker, but a more natural approach would use an `enum`.

jodh-intel commented on src/shim.c:
```diff
@@ -553,56 +616,46 @@ main(int argc, char **argv)
-	long long          val;
+	struct addrinfo    hints;
+	struct addrinfo   *servinfo, *addr;
+	char               port_str[20];
```
Magic number. Why `20`? I'd use `BUFSIZE` or use a `#define` explaining why 20 bytes is sufficient.

jodh-intel commented on src/shim.c:
```diff
@@ -553,56 +616,46 @@ main(int argc, char **argv)
-		{"debug", no_argument, 0, 'd'},
+		{"token", required_argument, 0, 't'},
+		{"proxy-address", required_argument, 0, 'a'},
+		{"proxy-port", required_argument, 0, 'p'},
```
Would be good to keep this list sorted.

jodh-intel commented on src/shim.c:
```diff
@@ -553,56 +616,46 @@ main(int argc, char **argv)
 	};
 
-	while ((c = getopt_long(argc, argv, "c:p:o:s:e:dh", prog_opts, NULL))!= -1) {
+	while ((c = getopt_long(argc, argv, "c:t:s:p:h", prog_opts, NULL))!= -1) {
```
This looks wrong - shouldn't it be `"a:c:p:t:h"`?

jodh-intel replied:
In fact, I think it should probably also include "`d`" - see comment above.

jodh-intel commented on src/shim.c:
```diff
@@ -553,56 +616,46 @@ main(int argc, char **argv)
-		{"seq-no", required_argument, 0, 's'},
-		{"err-seq-no", required_argument, 0, 'e'},
-		{"debug", no_argument, 0, 'd'},
+		{"token", required_argument, 0, 't'},
```
Did you mean to remove the debug option?

jodh-intel commented on src/shim.c:
```diff
@@ -620,29 +673,60 @@ main(int argc, char **argv)
+
+	sprintf(port_str, "%d", shim.proxy_port);
+	if ((ret = getaddrinfo(shim.proxy_address, port_str, &hints, &servinfo)) != 0) {
+		fprintf(stderr, "getaddrinfo error: %s\n", strerror(errno));
```
It would be useful to add `shim.proxy_address` to the error message.

jodh-intel commented on src/shim.c:
```diff
@@ -620,29 +673,60 @@ main(int argc, char **argv)
+	for(addr = servinfo; addr != NULL; addr = addr->ai_next) {
+		if ((sockfd = socket(addr->ai_family, addr->ai_socktype,
+			addr->ai_protocol)) == -1) {
+			shim_debug("Error in socket creation:%s\n", strerror(errno));
```
Again, this error could do with more details.

jodh-intel commented on src/shim.c:
```diff
@@ -620,29 +673,60 @@ main(int argc, char **argv)
+
+		if (connect(sockfd, addr->ai_addr, addr->ai_addrlen) == -1) {
+			close(sockfd);
+			shim_debug("Error in client connection:%s\n", strerror(errno));
```
Ditto.

jodh-intel commented on src/shim.c:
```diff
@@ -620,29 +673,60 @@ main(int argc, char **argv)
+	}
+
+ 	if (addr == NULL) {
+		shim_error("Failed to connect to proxy: %s\n", strerror(errno));
```
Which proxy? More detail required.

jodh-intel commented on src/shim.c:
```diff
@@ -620,29 +673,60 @@ main(int argc, char **argv)
+	/* Send a Connect command to the proxy and wait for the response
+	 */
+	if (! send_proxy_connect_cmd(&shim)) {
+		shim_error("Could not send connect command to proxy\n");
```
Which proxy? More detail required.

jodh-intel commented on src/utils.c:
```diff
@@ -97,6 +97,32 @@ set_fd_nonblocking(int fd)
+ * \param val Short Integer to be converted to big endian
+ */
+void
+set_big_endian_16(uint8_t *buf, uint16_t val)
```
I'd consider simply returning if buf is `NULL` (or making the function return a `bool`).

jodh-intel commented on src/utils.c:
```diff
@@ -97,6 +97,32 @@ set_fd_nonblocking(int fd)
+ * \return Unsigned 16 bit network ordered integer
+ */
+uint16_t
+get_big_endian_16(const uint8_t *buf)
```
Same comment as above about `buf`.

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+	msg = calloc((size_t)len, sizeof(uint8_t));
 
-	if (! proxy_ctl_msg) {
+	if (! msg) {
```
There are callers of this function which check the return value, but the value will always be true due to this check. Should it return `NULL` rather than calling `abort()` maybe?

amshinde replied:
I think its best to abort when we run into memory allocation failures. Agree this makes checking for NULL on the caller side redundant, I'd prefer to remove that check instead.

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+	set_big_endian_16(msg, PROXY_VERSION);
+	/* Header length expressed in 32 bit words.
+	 */
+	msg[HEADER_LEN_OFFSET] = (header->header_len/32);
```
Magic number. What happens when the protocol changes to some other value? I'd create a documented `#define` and use that.

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
-	ret = asprintf(&proxy_payload,
-			"{\"id\":\"%s\",\"data\":{\"hyperName\":\"%s\",\"data\":%s}}",
-			proxy_command_id, hyper_cmd, json);
+	total_size = (ssize_t)((header->header_len * 32) + header->payload_len);
```
Another place you could use that `#define`.

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+ * \param shim \ref cc_shim
+ * \param header \ref frame_header
+ * \param payload Payload to be sent
+ */
```
doxygen will complain if you run it on this code due to the missing `\return`. Same comment for various other functions below.

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+ */
+bool
+send_proxy_message(struct cc_shim *shim, uint8_t type, uint8_t opcode,
+			const char *payload)
```
Missing check on `payload`.

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+	char *payload = NULL;
+	bool ret;
+
+	if (! (shim && shim->token) || shim->proxy_sock_fd < 0) {
```
I'd break this check up to make it clearer that the check on the `shim` pointer is fully handled before the checks on the members:

```
if (! shim) {
    return false;
}

if ((! shim->token) || (shim->proxy_sock_fd < 0)) {
    return false;
}
```

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+	uint8_t *buf = NULL;
+	size_t  size = VERSION_SIZE + HEADER_LEN_SIZE;
+
+	if (! (shim && header)) {
```
No check on `payload`.

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+		return false;
+	}
+
+	buf  = calloc(size, sizeof(uint8_t));
```
No check on `buf`?

jodh-intel commented on src/shim.c:
```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
+
+	header->header_len = buf[HEADER_LEN_OFFSET];
+
+	buf = realloc(buf, (size_t)header->header_len * 32);
```
No check on `buf`?

jodh-intel replied:
Magic number again.

jodh-intel commented on README.md:
```diff
@@ -5,21 +5,20 @@
-
-TODO:
-The shim should capture the exit status of the container and exit with that exit code.
+writes any strean data received from the proxy to its stdout/stderr streams.
```
typo: "strean".

amshinde replied:
Fixed

jodh-intel commented on src/shim.h:
```diff
@@ -18,44 +18,70 @@
  * fd passed by the runtime
  */
-#define MAX_POLL_FDS 4
+#define MAX_POLL_FDS 3
```
Comment is now incorrect.

amshinde replied:
Fixed

jodh-intel commented on src/shim.h:
```diff
@@ -18,44 +18,70 @@
+	uint8_t *payload;
+};
+
+enum frametype {
```
Unused?

jodh-intel commented on src/shim.h:
```diff
@@ -18,44 +18,70 @@
+	cmd_signal,
+};
+
+enum streamtype {
```
Unused?

jodh-intel commented on src/shim.h:
```diff
@@ -18,44 +18,70 @@
+	stream_stderr,
+};
+
+enum notificationtype {
```
Unused?

jodh-intel commented on src/shim.h:
```diff
@@ -18,44 +18,70 @@
+	type_notification
+};
+
+enum commandtype {
```
Unused?

amshinde replied:
I have used the values in calls to send_proxy_message() and checking the frame header elements for equality after reading a frame.

jodh-intel replied:
Ah - right, so partially used. To make the code easier to read, a good convention is to use the name of the `enum` as a prefix to the values:

```
enum foo {
    foo_value_a,
    foo_value_b,
    foo_value_c,
};

```

jodh-intel commented on src/shim.h:
```diff
@@ -18,44 +18,70 @@
- */
-#define HYPERSTART_MAX_RECV_BYTES       10240
+// Header size is length of header in 32 bit words.
+#define  MIN_HEADER_SIZE    3
```
All other sizes are in bytes, so this could be the source of possible errors as the assumption will be that's 3 bytes. Can it be expressed in bytes for consistency instead or renamed to make it clear this is *not* in bytes?

amshinde replied:
Changed this to MIN_HEADER_WORD_SIZE

jodh-intel commented on src/shim.h:
```diff
@@ -18,44 +18,70 @@
- * source where hyper_event_ops->wbuf_size is set).
- */
-#define HYPERSTART_MAX_RECV_BYTES       10240
+// Header size is length of header in 32 bit words.
```
It's a shame to lose that ascii art as it makes all these defines more meaningful imho.

jodh-intel commented on src/shim.c:
```diff
@@ -322,188 +505,133 @@ handle_stdin(struct cc_shim *shim)
-						(unsigned long int)stream_len);
-				goto err;
-			}
+	if (fr->header.opcode == stream_stdout) {
```
Since there are multiple opcodes with the same value, I'd perform a check on `fr->header.type` at the top of this function to assert that you are actually operating on a `type_stream` frame.

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,278 @@ err_exit(const char *format, ...)
+ * \return true on success, false otherwise
+ */
+bool
+send_proxy_message(struct cc_shim *shim, uint8_t type, uint8_t opcode,
```
This might be a good place to add a log call to make debugging possible.

jodh-intel commented on src/shim.c:
```diff
@@ -322,188 +505,133 @@ handle_stdin(struct cc_shim *shim)
+			break;
+		default:
+			shim_warning("Unknown frame with type %d received\n",
+				fr->header.type);
```
Nice. Again though, might be good to add a log call here to make debugging possible.

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,283 @@ err_exit(const char *format, ...)
+
+	// TODO: Verify the payload format
+	ret = asprintf(&payload,
+			"{\"token\":\"%s\"}", shim->token);
```
What if `shim->token` containers a double-quote? We should sanitise/validate the values passed via `--token`.

amshinde replied:
We should really expect the proxy/runtime to send the token correctly. The shim should just forward this token as is.

jodh-intel replied:
That's an unsafe assumption: `shim->token` is a value passed in from outside this program, so needs to be validated.

amshinde replied:
@jodh-intel I have added a commit to get rid of any quotes if present in the token.

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,283 @@ err_exit(const char *format, ...)
+			// to proxy.
+			exit(EXIT_FAILURE);
+		} else if (ret < 0) {
+			shim_error("Failed to talk to proxy: %s\n", 
```
Which proxy? Maybe display some details to aid debugging failure scenarios.

amshinde replied:
Same as above

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,283 @@ err_exit(const char *format, ...)
+	do {
+		ret = recv(fd, buf, (size_t)(size-offset), 0);
+		if (ret == 0) {
+			shim_error("Proxy has exited\n");
```
Which proxy?

amshinde replied:
This a function that works on an fd, intended for writing tests. Will change the error messages accordingly.

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,283 @@ err_exit(const char *format, ...)
+ * \param shim \ref cc_shim
+ * \param header \ref frame
+ *
+ * \return true on success, false otherwise
```
This is incorrect.

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,283 @@ err_exit(const char *format, ...)
+	/* Get the header size in bytes, as the proxy passes the header length
+	 * as 32 bit word length.
+	 */
+	header_size_in_bytes = (size_t)fr->header.header_len * 4;
```
`sizeof(int32_t)` rather than `4` makes the intent clear.

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,283 @@ err_exit(const char *format, ...)
+	}
+
+	if (! read_wire_data(shim->proxy_sock_fd, buf, size)) {
+		return false;
```
This looks like a leak: `fr` and `buf` are not freed.

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,283 @@ err_exit(const char *format, ...)
+	fr->header.version = get_big_endian_16(buf);
+
+	if (fr->header.version != PROXY_VERSION) {
+		shim_error("Bad frame version: %d\n", fr->header.version);
```
I'd also display the expected version.

jodh-intel commented on src/shim.c:
```diff
@@ -322,188 +510,148 @@ handle_stdin(struct cc_shim *shim)
-		}
+	// Reponses received from proxy are currently just logged.
+	if (fr->header.err) {
+		shim_error("Error response received from proxy: %s\n", fr->payload);
```
Which proxy?

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -322,188 +510,148 @@ handle_stdin(struct cc_shim *shim)
+		 * to requests in future. Maybe useful for restarting connection with proxy
+		 * and resending requests that were missed.
+		 */
+		shim_debug("Response received from proxy: %s\n", fr->payload);
```
Which proxy?

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -322,188 +510,148 @@ handle_stdin(struct cc_shim *shim)
+				 (size_t)(fr->header.payload_len - offset));
 		if (ret <= 0 ) {
-			goto out;
+			shim_error("Could not write stream to fd %d : %s\n",
```
Which proxy?

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -528,18 +676,193 @@ parse_numeric_option(char *input) {
+parse_connection_uri(struct cc_shim *shim, char *uri)
+{
+	char    *unix_uri = "unix://";
+	char    *tcp_uri = "tcp://";
```
`const` for both these?

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -528,18 +676,193 @@ parse_numeric_option(char *input) {
+	}
+
+	if (! strncmp(uri, unix_uri, unix_uri_len)) {
+		shim->proxy_address = strdup(uri + strlen(unix_uri));
```
You've already calculated `strlen(unix_uri)` for `unix_uri_len`.

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -528,18 +676,193 @@ parse_numeric_option(char *input) {
+
+		shim->proxy_port = (int)parse_numeric_option(port_offset + 1);
+		if (shim->proxy_port == -1) {
+			goto out;
```
Maybe add an debug call here showing what could not be parsed as a port?

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -528,18 +676,193 @@ parse_numeric_option(char *input) {
+		addr_len = port_offset - (uri + tcp_uri_len);
+
+		if (addr_len == 0) {
+			shim_error("Missing tcp hostname in uri\n");
```
Display full uri?

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -528,18 +676,193 @@ parse_numeric_option(char *input) {
+
+		if (connect(sockfd, (struct sockaddr *)&remote,
+				sizeof(struct sockaddr_un)) == -1) {
+			shim_error("Error while connecting to proxy"\
```
Backslash isn't needed.

amshinde replied:
Fixed

jodh-intel commented on src/shim.c:
```diff
@@ -528,18 +676,193 @@ parse_numeric_option(char *input) {
+	shim->proxy_sock_fd = sockfd;
+	return true;
+out:
+	free(port_str);
```
`port_str` is only set if `shim->proxy_port != -1`.

amshinde replied:
I have initialized it to NULL, so free(NULL) should be a nop.

dlespiau commented on src/shim.h:
```diff
@@ -27,6 +27,22 @@ struct cc_shim {
 	int         proxy_port;
 };
 
+/*  0 1 2 3 4 5 7 0 1 2 3 4 5 6 7 0 1 2 3 4 5 6 7 0 1 2 3 4 5 6 7
```
you're missing one line above that one with:

```
                    1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
```

amshinde replied:
What does that stand for?

grahamwhaley commented on src/shim.h:
```diff
@@ -15,47 +15,88 @@
- * | ctrl id | length  | payload (length-8)      |
- * | . . . . | . . . . | . . . . . . . . . . . . |
- * 0         4         8                         length
+/*  0 1 2 3 4 5 7 0 1 2 3 4 5 6 7 0 1 2 3 4 5 6 7 0 1 2 3 4 5 6 7
```
err, is there a '6' missing in that first block of digits in the comment, or am I missing something? :-)

amshinde replied:
yup ! Adding it.

jodh-intel commented on src/shim.c:
```diff
@@ -151,97 +155,287 @@ err_exit(const char *format, ...)
+		return false;
+	}
+
+	strip_double_quotes(shim->token);
```
I think it would be better to just fail if `shim->token` contains any non-alphanumerics. That admittedly does mean the shim would then have knowledge of the token format (rather than just passing the data through), but it would be a quick and safe approach that only lets through data in the expected format.

dlespiau replied:
fwiw token is specified to be base64 encoded (RFC 4648, using `-` and `_` and `=` for padding). The token themselves are 32 bytes today.

jodh-intel replied:
Thanks for clarifying @dlespiau.

amshinde replied:
@jodh-intel @dlespiau thanks, I'll add a check to verify that the token contains characters permitted by base64 encoding

amshinde replied:
@jodh-intel I have added a check to verify that the token passed is a valid base64url encoded string. Have a look. Will be pushing a functional test in my other PR for tests to test this.

dlespiau commented on src/shim.c:
```diff
@@ -528,6 +533,186 @@ parse_numeric_option(char *input) {
+{
+	const char *unix_uri = "unix://";
+	const char *tcp_uri = "tcp://";
+	size_t      unix_uri_len = strlen(unix_uri);
```
This could be hardened a bit with strnlen() (PATH_MAX?)

dlespiau commented on src/shim.c:
```diff
@@ -528,6 +533,186 @@ parse_numeric_option(char *input) {
+		return false;
+	}
+
+	if (! strncmp(uri, unix_uri, unix_uri_len)) {
```
or PATH_MAX directly here.

amshinde commented:
Rebased changes to merge conflicting travis file with master. 
Added 4 commits, after the last review.
@jodh-intel @dlespiau @sboeuf  Can you review the last 4 commits, so that this can be merged?

jodh-intel commented on src/shim.c:
```diff
@@ -631,8 +652,13 @@ handle_proxy_notification(struct cc_shim *shim, struct frame *fr)
+		if (! send_proxy_message(shim, frametype_command,
+					cmd_disconnectshim, NULL)) {
+			shim_error("Could not send Disconnect shim command"
+				"to proxy at %s\n", shim->proxy_address);
```
Missing space between "command" and "to".

jodh-intel commented on src/shim.c:
```diff
@@ -434,6 +441,13 @@ read_frame(struct cc_shim *shim)
 	}
+
+	shim_debug("Read frame HeaderLength: %d, Version: %d, Payloadlen: %d,"
+			"Type:%d, Opcode: %d, Err:%d\n",
```
Missing space before "Type".

jodh-intel commented on src/shim.c:
```diff
@@ -434,6 +441,13 @@ read_frame(struct cc_shim *shim)
 		free(buf);
 	}
+
+	shim_debug("Read frame HeaderLength: %d, Version: %d, Payloadlen: %d,"
```
It would be easier to read this if you broke the format string so into chunks containing only 2 formats as that would align with the way you've paired up the variables below.

jodh-intel commented on src/shim.c:
```diff
@@ -312,6 +312,11 @@ send_connect_command(struct cc_shim *shim)
 	ret = send_proxy_message(shim, frametype_command, cmd_connectshim, payload);
+	if (! ret) {
+		shim_error("Could not send initial connect command to"
+				"proxy at %s\n", shim->proxy_address);
```
Missing space between "to" and "proxy".

amshinde commented:
@jodh-intel I have addressed your review comments. Please have a look.

jodh-intel commented on src/utils.c:
```diff
@@ -96,6 +98,87 @@ set_fd_nonblocking(int fd)
+		return -1;
+	}
+
+	ret = asprintf(&regex, "^(%1$s{4})*(%1$s{4}|%1$s{3}=|%1$s{2}==)$",
```
Let's hope we never need to touch this line! It's written in Malbolge right? :smile:

amshinde replied:
Ah.. I had to look that up! Its just good old regular expressions :)

amshinde commented:
@dlespiau @sboeuf @sameo @sameo Can we have this merged?

sboeuf commented:
LGTM

sameo commented:
LGTM
